src/connectors/voice-search/__tests__/connectVoiceSearch-test.js

```diff
@@ -0,0 +1,188 @@
+import jsHelper, { SearchParameters } from 'algoliasearch-helper';
+import connectVoiceSearch from '../connectVoiceSearch';
+
+jest.mock('../../../lib/voiceSearchHelper', () => {
+  return ({ onStateChange, onQueryChange }) => {
+    return {
+      getState: () => {},
+      isBrowserSupported: () => true,
+      isListening: () => false,
+      toggleListening: () => {},
+      // ⬇️ for test
+      changeState: () => onStateChange(),
+      changeQuery: query => onQueryChange(query),
+    };
+  };
+});
+
+function getDefaultSetup() {
+  const renderFn = jest.fn();
+  const unmountFn = jest.fn();
+  const makeWidget = connectVoiceSearch(renderFn, unmountFn);
+  const widget = makeWidget({});
+  const helper = jsHelper({});
+
+  return {
+    renderFn,
+    unmountFn,
+    widget,
+    helper,
+  };
+}
+
+function getInitializedWidget() {
+  const { renderFn, unmountFn, widget, helper } = getDefaultSetup();
+
+  helper.search = () => {};
+  widget.init({ helper });
+
+  return {
+    renderFn,
+    unmountFn,
+    widget,
+    helper,
+    refine: query => widget._refine(query),
+  };
+}
+
+describe('connectVoiceSearch', () => {
+  describe('Usage', () => {
+    it('throws without render function', () => {
+      expect(() => {
+        connectVoiceSearch()({});
+      }).toThrowErrorMatchingInlineSnapshot(`
+"The render function is not valid (got type \\"undefined\\").
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/voice-search/js/#connector"
+`);
+    });
+  });
+
+  describe('lifecycle', () => {
+    it('calls renderFn during init and render', () => {
+      const { renderFn, widget, helper } = getDefaultSetup();
+      widget.init({ helper });
+      expect(renderFn).toHaveBeenCalled();
+      expect(renderFn).toHaveBeenLastCalledWith(
+        expect.objectContaining({}),
+        true
+      );
+      widget.render({
+        helper,
+      });
+      expect(renderFn).toHaveBeenCalledTimes(2);
+      expect(renderFn).toHaveBeenLastCalledWith(
+        expect.objectContaining({}),
+        false
+      );
+    });
+
+    it('calls unmount on dispose', () => {
+      const { unmountFn, widget, helper } = getDefaultSetup();
+      widget.init({ helper });
+      widget.dispose({ helper, state: helper.getState() });
+      expect(unmountFn).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  it('render triggered when state changes', () => {
+    const { renderFn, widget, helper } = getDefaultSetup();
+    widget.init({ helper });
+    expect(renderFn).toHaveBeenCalled();
+    widget._voiceSearchHelper.changeState();
```

Comment:
It'd be nice not to touch the internals in our tests. Not sure how to achieve that here.

Reply:
```
                                    (A)                             (B)
voiceSearchHelper internal logics  ----> `onStateChange` triggered ----> render triggered
```

I want to test (B) and now the voiceSearchHelper is mocked. So I had to fake (A). I don't know if it's possible without touching the internals of voiceSearchHelper.